File: mobile/gps-tracker-ios/GpsTracker/Services/SpeedGradient.swift
import CoreLocation
import UIKit

/// Speed → color stops shared with the web demo (`web/gpx-track-demo`).
/// Low range is stretched so 5 / 15 / 25 km/h read as different colors.
enum SpeedGradient {
    struct Stop {
        let kmh: Double
        let color: UIColor
    }

    /// Stops (km/h): walk / slow transit contrast first, then car/MTB tail.
    static let stops: [Stop] = [
        Stop(kmh: 0, color: UIColor(red: 0.086, green: 0.639, blue: 0.290, alpha: 1)), // #16a34a
        Stop(kmh: 5, color: UIColor(red: 0.133, green: 0.773, blue: 0.369, alpha: 1)), // #22c55e
        Stop(kmh: 15, color: UIColor(red: 0.635, green: 0.839, blue: 0.133, alpha: 1)), // #a2d622
        Stop(kmh: 25, color: UIColor(red: 0.918, green: 0.702, blue: 0.031, alpha: 1)), // #eab308
        Stop(kmh: 50, color: UIColor(red: 0.976, green: 0.451, blue: 0.086, alpha: 1)), // #f97316
        Stop(kmh: 90, color: UIColor(red: 0.937, green: 0.267, blue: 0.267, alpha: 1)), // #ef4444
        Stop(kmh: 140, color: UIColor(red: 0.659, green: 0.333, blue: 0.969, alpha: 1)), // #a855f7
        Stop(kmh: 200, color: .black),
    ]

    /// Display-only: ignore tiny Δt (duplicate fixes).
    static let minSegmentDurationS: TimeInterval = 0.35
    /// Rolling median half-window (total width = 2*r+1).
    static let medianRadius = 2
    /// Max |Δv| between consecutive display samples (m/s²). ~human/transit jerk.
    static let maxAccelerationMps2: Double = 3.5

    static func color(kmh: Double) -> UIColor {
        let v = max(0, kmh)
        if v <= stops[0].kmh { return stops[0].color }
        for i in 1..<stops.count {
            let a = stops[i - 1]
            let b = stops[i]
            if v <= b.kmh {
                let t = (v - a.kmh) / (b.kmh - a.kmh)
                return lerp(a.color, b.color, t: t)
            }
        }
        return stops.last!.color
    }

    /// Instantaneous Δs/Δt (km/h). Gaps without usable time → 0.
    static func segmentSpeedKmh(from: CLLocation, to: CLLocation) -> Double {
        let dt = to.timestamp.timeIntervalSince(from.timestamp)
        guard dt >= minSegmentDurationS else { return 0 }
        let meters = to.distance(from: from)
        return (meters / dt) * 3.6
    }

    /// Speeds for map coloring: median + acceleration clamp.
    /// Does not mutate recorded GPX — display only.
    static func displaySpeedsKmh(for locations: [CLLocation]) -> [Double] {
        let raw = speedsKmh(for: locations)
        guard !raw.isEmpty else { return [] }

        let medians = rollingMedian(raw, radius: medianRadius)
        return clampAcceleration(medians, locations: locations)
    }

    static func speedsKmh(for locations: [CLLocation]) -> [Double] {
        guard locations.count >= 2 else { return [] }
        var out: [Double] = []
        out.reserveCapacity(locations.count - 1)
        for i in 1..<locations.count {
            // Prefer CoreLocation's filtered speed on the newer fix when present.
            let device = locations[i].speed
            if device >= 0 {
                out.append(device * 3.6)
            } else {
                out.append(segmentSpeedKmh(from: locations[i - 1], to: locations[i]))
            }
        }
        return out
    }

    private static func rollingMedian(_ values: [Double], radius: Int) -> [Double] {
        guard radius > 0, values.count > 1 else { return values }
        return values.indices.map { i in
            let lo = max(0, i - radius)
            let hi = min(values.count - 1, i + radius)
            let window = Array(values[lo...hi]).sorted()
            return window[window.count / 2]
        }
    }

    private static func clampAcceleration(_ speedsKmh: [Double], locations: [CLLocation]) -> [Double] {
        guard speedsKmh.count >= 2 else { return speedsKmh }
        var out = speedsKmh
        for i in 1..<out.count {
            let dt = locations[i + 1].timestamp.timeIntervalSince(locations[i].timestamp)
            guard dt >= minSegmentDurationS else {
                out[i] = out[i - 1]
                continue
            }
            let prev = out[i - 1] / 3.6
            let curr = out[i] / 3.6
            let maxDelta = maxAccelerationMps2 * dt
            let clamped = min(max(curr, prev - maxDelta), prev + maxDelta)
            out[i] = max(0, clamped * 3.6)
        }
        return out
    }

    private static func lerp(_ a: UIColor, _ b: UIColor, t: Double) -> UIColor {
        var ar: CGFloat = 0, ag: CGFloat = 0, ab: CGFloat = 0, aa: CGFloat = 0
        var br: CGFloat = 0, bg: CGFloat = 0, bb: CGFloat = 0, ba: CGFloat = 0
        a.getRed(&ar, green: &ag, blue: &ab, alpha: &aa)
        b.getRed(&br, green: &bg, blue: &bb, alpha: &ba)
        let u = CGFloat(max(0, min(1, t)))
        return UIColor(
            red: ar + (br - ar) * u,
            green: ag + (bg - ag) * u,
            blue: ab + (bb - ab) * u,
            alpha: aa + (ba - aa) * u
        )
    }
}

File: mobile/gps-tracker-android/app/src/main/kotlin/com/vlandivir/gpstracker/gpx/SpeedGradient.kt
package com.vlandivir.gpstracker.gpx

import kotlin.math.max
import kotlin.math.min

/** Speed → color stops shared with `/gpx-track-demo` and the iOS app. Packed ARGB. */
object SpeedGradient {
    data class Stop(val kmh: Double, val color: Int)

    val stops: List<Stop> = listOf(
        Stop(0.0, rgb(0x16, 0xa3, 0x4a)),
        Stop(5.0, rgb(0x22, 0xc5, 0x5e)),
        Stop(15.0, rgb(0xa2, 0xd6, 0x22)),
        Stop(25.0, rgb(0xea, 0xb3, 0x08)),
        Stop(50.0, rgb(0xf9, 0x73, 0x16)),
        Stop(90.0, rgb(0xef, 0x44, 0x44)),
        Stop(140.0, rgb(0xa8, 0x55, 0xf7)),
        Stop(200.0, rgb(0x00, 0x00, 0x00)),
    )

    const val minSegmentDurationS = 0.35
    const val medianRadius = 2
    const val maxAccelerationMps2 = 3.5

    fun color(kmh: Double): Int {
        val v = max(0.0, kmh)
        if (v <= stops.first().kmh) return stops.first().color
        for (i in 1 until stops.size) {
            val a = stops[i - 1]
            val b = stops[i]
            if (v <= b.kmh) {
                val t = (v - a.kmh) / (b.kmh - a.kmh)
                return lerp(a.color, b.color, t)
            }
        }
        return stops.last().color
    }

    fun segmentSpeedKmh(from: GeoFix, to: GeoFix): Double {
        val dt = (to.timestampMs - from.timestampMs) / 1000.0
        if (dt < minSegmentDurationS) return 0.0
        return (to.distanceTo(from) / dt) * 3.6
    }

    fun displaySpeedsKmh(locations: List<GeoFix>): List<Double> {
        val raw = speedsKmh(locations)
        if (raw.isEmpty()) return emptyList()
        return clampAcceleration(rollingMedian(raw, medianRadius), locations)
    }

    fun speedsKmh(locations: List<GeoFix>): List<Double> {
        if (locations.size < 2) return emptyList()
        return (1 until locations.size).map { i ->
            val device = locations[i].speedMps
            if (device != null && device >= 0) device * 3.6
            else segmentSpeedKmh(locations[i - 1], locations[i])
        }
    }

    private fun rollingMedian(values: List<Double>, radius: Int): List<Double> {
        if (radius <= 0 || values.size <= 1) return values
        return values.indices.map { i ->
            val lo = max(0, i - radius)
            val hi = min(values.lastIndex, i + radius)
            val window = values.subList(lo, hi + 1).sorted()
            window[window.size / 2]
        }
    }

    private fun clampAcceleration(speedsKmh: List<Double>, locations: List<GeoFix>): List<Double> {
        if (speedsKmh.size < 2) return speedsKmh
        val out = speedsKmh.toMutableList()
        for (i in 1 until out.size) {
            val dt = (locations[i + 1].timestampMs - locations[i].timestampMs) / 1000.0
            if (dt < minSegmentDurationS) {
                out[i] = out[i - 1]
                continue
            }
            val prev = out[i - 1] / 3.6
            val curr = out[i] / 3.6
            val maxDelta = maxAccelerationMps2 * dt
            val clamped = min(max(curr, prev - maxDelta), prev + maxDelta)
            out[i] = max(0.0, clamped * 3.6)
        }
        return out
    }

    fun rgb(r: Int, g: Int, b: Int): Int =
        (0xFF shl 24) or (r shl 16) or (g shl 8) or b

    private fun lerp(a: Int, b: Int, t: Double): Int {
        val u = t.coerceIn(0.0, 1.0)
        fun chan(shift: Int): Int {
            val av = (a shr shift) and 0xFF
            val bv = (b shr shift) and 0xFF
            return (av + (bv - av) * u).toInt()
        }
        return (chan(24) shl 24) or (chan(16) shl 16) or (chan(8) shl 8) or chan(0)
    }
}
